### ISSUE TITLE
FormatException when passing interpolated string containing '{' to Console.WriteLine

### ISSUE THREAD

**Issue by junderhill** [dotnet/core#5153]
I've seen an issue recently when using a logging class which just uses `Console.WriteLine` when we call the Logging methods with an interpolated string containing serialised JSON. After some investigation it seems that if the interpolated string contains a curly brace that when this string is used in `Console.WriteLine` with no params then a FormatException is thrown.

The code sample below may explain this better.

As a work around we can remove the `params object[] args` from our logging class for now and go back and change where we were using this to only use string interpolation but it definitely seems like a bug?

# Sample Code

```
static void Main(string[] args)
        {
          var myString = "{}";
          LogNoArgs($"Data: {myString}"); //this works
          Log($"Data: {myString}"); //this throws FormatException when WriteLine is called in the Log Method


          var myString2 = "";
          Log($"Data: {myString2}"); //this works
        }

        public static void Log(string message, params object[] args){
          Console.WriteLine(message, args);
        }
        public static void LogNoArgs(string message){
          Console.WriteLine(message);
        }

```

# System Info

```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.300
 Commit:    b2475c1295

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.15
 OS Platform: Darwin
 RID:         osx.10.15-x64
 Base Path:   /usr/local/share/dotnet/sdk/3.1.300/

Host (useful for support):
  Version: 3.1.4
  Commit:  0c2e69caa6

.NET Core SDKs installed:
  3.1.101 [/usr/local/share/dotnet/sdk]
  3.1.300 [/usr/local/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.1 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.4 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.1 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.4 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
```

**Comment by omajid** [comment]
Unless I misunderstood the issue, this is by design.

The `Console.WriteLine(string, object[])` call that breaks is basically this:

```
    public static void Main() {
        Console.WriteLine("{}", new object[0]);
    }
```

[See in CSharpLab](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUHjAZgAIBvPYy4jdMiqh1TATgAoAiUgX3ZmKgCmAd2IB7AEYArAQGNgAbQAMAXQCUAbnqUueLkA=)

This also throws a `FormatException`:

```
System.FormatException: Input string was not in a correct format.
   at System.Text.StringBuilder.FormatError()
   at System.Text.StringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.FormatHelper(IFormatProvider provider, String format, ParamsArray args)
   at System.String.Format(IFormatProvider provider, String format, Object[] args)
   at System.IO.TextWriter.WriteLine(String format, Object[] arg)
   at System.IO.TextWriter.SyncTextWriter.WriteLine(String format, Object[] arg)
   at System.Console.WriteLine(String format, Object[] arg)
   at Program.Main()
```

The [`Console.WriteLine(string, object[])` docs](https://docs.microsoft.com/en-us/dotnet/api/system.console.writeline?view=netcore-3.1#System_Console_WriteLine_System_String_System_Object___) describe this feature:

> This method uses the composite formatting feature of the .NET Framework to convert the value of an object to its text representation and embed that representation in a string. The resulting string is written to the output stream.
>
> The format parameter consists of zero or more runs of text intermixed with zero or more indexed placeholders, called format items, that correspond to an object in the parameter list of this method. The formatting process replaces each format item with the text representation of the value of the corresponding object.
>
> The syntax of a format item is {index[,alignment][:formatString]}, which specifies a mandatory index, the optional length and alignment of the formatted text, and an optional string of format specifier characters that govern how the value of the corresponding object is formatted.

In other words, `Console.WriteLine(string, object[])` treats text in parenthesis (`{...}`) in the input string as a special case and tries to format it using the supplied `object[]`. If you have have a string that needs formatting, you should escape the empty braces `{}` or `{ }`. If you dont have a string that needs formatting, you should use the non-formatting variant `Console.WriteLine(string)`.

IMO, this is a feature of `Console.WriteLine(string, object[]`). It's not a bug and unlikely to be changed.

**Comment by junderhill** [comment]
I see what you’re saying but I think perhaps I have you a bad example. 

Doing the same but with a different string also causes the issue, I wouldn’t expect that to fail.

var myString = "{‘message’:’hello’}";
          LogNoArgs($"Data: {myString}"); //this works
          Log($"Data: {myString}")

**Comment by omajid** [comment]
Sorry if I misunderstood something. Your new example boils down to:

```
var myString = "{‘message’:’hello’}";
Console.WriteLine($"Data: {myString}"); //this works
Console.WriteLine($"Data: {myString}", new object[0]);
```

The string interpolation is done at compile-time. The generated C#, compiled back to C#, is simply:

```
        string str = "{‘message’:’hello’}";
        Console.WriteLine("Data: " + str);
        Console.WriteLine("Data: " + str, new object[0]);
```

`Console.WriteLine(string)` takes [a single string and doesn't do any formatting](https://docs.microsoft.com/en-us/dotnet/api/system.console.writeline?view=netcore-3.1#System_Console_WriteLine_System_String_). So it works, no matter what argument you give it.

 `Console.WriteLine(string, object[])` takes [an array of arguments and uses it to format any text enclosed in parenthesis (`{...}`) in the input string](https://docs.microsoft.com/en-us/dotnet/api/system.console.writeline?view=netcore-3.1#System_Console_WriteLine_System_String_System_Object___). It doesn't work if you give it format strings that it isn't expecting (such as valid json). You need to [escape these braces](https://docs.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting#escaping-braces).
